### DISCUSSION TITLE
Use ebitenmobile to build for android throws errors (ebitenmobileview_android.c)

### DISCUSSION THREAD

**Discussion by scratch85** [hajimehoshi/ebiten#2893] [Q&A] [answered]
Hi,

I wanted to build for android and adopted my project to look like the go-inovation example. Sadly I receive an error where I am not sure what could be the cause or how to debug further.

```
go version
go version go1.21.6 windows/amd64
```

From go.mod
```
module client

go 1.20

require github.com/hajimehoshi/ebiten/v2 v2.6.4
```

Error using ebitenmobile
```
PS> ebitenmobile bind -target android -javapkg com.warlords-client.game -o ./mobile/android/client/client.aar -androidapi=21 ./mobile
gomobile: go build -buildmode=c-shared -o=gomobile-work-680967744\android\src\main\jniLibs\x86_64\libgojni.so ./gobind failed: exit status 1
# gobind/gobind
ebitenmobileview_android.c:22:1: error: variable has incomplete type 'void'
ebitenmobileview_android.c:22:20: error: expected ';' after top level declarator
```

Any help is appreciated.

Thanks!

**Comment by hajimehoshi** [comment]
Which version of ebitenmobile are you using?

**Reply by hajimehoshi** [reply]
`ebitenmobileview_android.c` is not a file in Ebitengine itself. Is this a file in your project?

**Reply by scratch85** [reply]
> Which version of ebitenmobile are you using?
I used `go get github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile@latest` so I guess it is using `v2@v2.7.0-alpha.5`, any chance to confirm this?!

`ebitenmobile` does not support something like `-v`, `--version` or `version`

> ebitenmobileview_android.c is not a file in Ebitengine itself. Is this a file in your project?

Actually it is not in my project. I am using pure go without doing any c code.

Also I checked [ebitenui](https://github.com/ebitenui/ebitenui), which I am using, but it is also not containing any *.c files.

**Reply by hajimehoshi** [reply]
> ebitenmobile does not support something like -v, --version or version

Which version of Ebitengine did you use to install ebitenmobile?

**Reply by hajimehoshi** [reply]
> I used go get github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile@latest

OK then you are using v2.6.4 for ebitenmobile.

**Reply by hajimehoshi** [reply]
Could you try to run ebitenmobile command with `-x` so that you can see detailed logs?

**Reply by scratch85** [reply]
> Could you try to run ebitenmobile command with -x so that you can see detailed logs?

Sure, here it is: [ebitenmobile-output-20240124.txt](https://github.com/hajimehoshi/ebiten/files/14042959/ebitenmobile-output-20240124.txt)
Already in line 27 it says: 
`gomobile: go build -x -buildmode=c-shared -o=C:\Users\me\AppData\Local\Temp\gomobile-work-3405399574\android\src\main\jniLibs\armeabi-v7a\libgojni.so ./gobind failed: exit status 1`

It looks like there is an issue with gobind which actually produces this file named `ebitenmobileview_android.c`

**Reply by hajimehoshi** [reply]
Thanks! Would it be possible to see the content of `ebitenmobileview_android.c`? You can keep the files by specifying `-work`.

**Reply by scratch85** [reply, marked as answer]
Ahh, this is something that helps a lot:

From the [gomobile docs](https://pkg.go.dev/golang.org/x/mobile/cmd/gomobile#hdr-Build_a_library_for_Android_and_iOS):
> The build flags -a, -i, -n, -x, -gcflags, -ldflags, -tags, -trimpath, and -work are shared with the build command. For documentation, see 'go help build'.

But hey, I found the issue today.
Using `-javapkg com.client.game` it works. Using `-javapkg com.warlords-client.game` fails with the error mentioned above.
Using `-javapkg com.warlordsclient.game` works as well, so the issue seems/is the "-" (minus/hyphen) in the `-javapkg` param or in general the way the cmd params (with go flag) work.

Many thanks for your help and your fast responses.

**Reply by hajimehoshi** [reply]
That's interesting. So using a hyphen didn't work well. I feel like this is a bug in gobind itself. I'd be nice if you could report this to github.com/golang/go.

**Reply by scratch85** [reply]
Just made a quick research about java package names. I think this is the root issue: Java does not support hyphens in package names.

- [Hyphenated company name in Java packages](https://stackoverflow.com/questions/3678520/hyphenated-company-name-in-java-packages)
- [Discussion on: Publishing a Python Package: What I Wish the Maze of Tutorials Covered](https://dev.to/cicirello/comment/1j0bl#:~:text=Java%20package%20names%20and%20module%20names%20can%20use%20underscores%20but%20not%20hyphens.)

Actually this means it is not a real bug, just some input that isn't handled well :)

**Reply by hajimehoshi** [reply]
ebitenmobile could detect this issue earlier. Let me file an issue.
